interpreter/hotspot/instance.go

```diff
@@ -453,42 +441,127 @@ func (d *hotspotInstance) getJITInfo(addr libpf.Address,
 		return nil, errors.New("JIT info evicted since eBPF snapshot")
 	}
 
+	method, err := d.getMethod(npsr.Ptr(nmethod, vms.Nmethod.Method), 0)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get JIT Method: %v", err)
+	}
+
 	// Finally read the associated debug information for this method
-	var scopesOff libpf.Address
-	metadataOff := npsr.PtrDiff32(nmethod, vms.Nmethod.MetadataOffset)
-	if vms.CompiledMethod.ScopesDataBegin != 0 {
-		scopesOff = npsr.Ptr(nmethod, vms.CompiledMethod.ScopesDataBegin) - addr
+	var jit *hotspotJITInfo
+	if vmd.version < 0x17000000 {
+		// JDK22 and earlier
+		//
+		// Layout of important bits in such 'class nmethod' pointer is:
+		// [class CodeBlob fields]
+		// [class CompiledMethod fields]
+		// [class nmethod fields]
+		// ...
+		// [JIT_code]		@ this + CodeBlob._code_start
+		// ...
+		// [metadata]		@ this + nmethod._metadata_offset	\ these three
+		// [scopes_data]	@ CompiledMethod._scopes_data_begin	| arrays we need
+		// [scopes_pcs]		@ this + nmethod._scopes_pcs_offset	/ for inlining info
+		// [dependencies]	@ this + nmethod._dependencies_offset
+		// ...
+		var scopesDataOff libpf.Address
+		metadataOff := npsr.PtrDiff32(nmethod, vms.Nmethod.MetadataOffset)
+		if vmd.nmethodUsesOffsets != 0 {
+			scopesDataOff = npsr.PtrDiff32(nmethod, vms.Nmethod.ScopesDataOffset)
+		} else {
+			scopesDataOff = npsr.Ptr(nmethod, vms.Nmethod.ScopesDataOffset) - addr
+		}
+		scopesPcsOff := npsr.PtrDiff32(nmethod, vms.Nmethod.ScopesPcsOffset)
+		depsOff := npsr.PtrDiff32(nmethod, vms.Nmethod.DependenciesOffset)
+
+		if metadataOff > scopesDataOff || scopesDataOff > scopesPcsOff || scopesPcsOff > depsOff {
+			return nil, fmt.Errorf("unexpected nmethod layout: %v <= %v <= %v <= %v",
+				metadataOff, scopesDataOff, scopesPcsOff, depsOff)
+		}
+
+		scopesData := make([]byte, depsOff-metadataOff)
```

Comment:
It it worth adding a maximum size check for `depsOff-metadataOff` here or is `addrCheck` / previous checks sufficient as a guard?

Reply:
This is moved/existing code. But since the JDK23 support checks for a 4MB limit, it probably should be added for symmetry here. I'll add a check for this.

Reply:
Done in commit e049ad7.